bot/cogs/antimalware.py

```diff
@@ -0,0 +1,43 @@
+import logging
+
+from discord import Message, utils
+from discord.ext.commands import Bot, Cog
+
+from bot.constants import AntiMalware as AntiMalwareConfig, Channels
+
+log = logging.getLogger(__name__)
+
+
+class AntiMalware(Cog):
+    """Cog providing anti-malware behavior."""
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.whitelist = tuple(AntiMalwareConfig.whitelist)
+
+    @Cog.listener()
+    async def on_message(self, message: Message) -> None:
+        """Identify messages with prohibited attachments."""
+        log.trace("Entered AntiMalware.on_message()")
+        rejected_attachments = [a for a in message.attachments if
+                                not a.filename.lower().endswith(self.whitelist)]
+        detected_pyfile = len([a for a in message.attachments if a.filename.lower().endswith('.py')]) > 0
```

Comment:
There is some code redundancy here.  I would rather make only on `for`.
```
for attachement in message.attachments:
      #if python_file add to detected_pyfile

      #if not in whitelist add to rejected_attachments
```

Are you sure one message can have several attachements?

Reply:
Testing seems to suggest that you're correct, they cannot. I misunderstood. Thanks!

Reply:
Messages can definitely have multiple attachments. Though the suggestion of using a single for loop is still a good suggestion.

Reply:
Okay. Can you help me understand why? Not trying to be argumentative, just trying to improve. Is the single loop preferred because iterating twice is simply more expensive? I considered that but thought the readability was better in this case, and that the size of the lists being so small would make it a trivial performance hit. 

Is this more like what you'd want to see instead?
```python
rejected_attachments = list()
detected_pyfile = list()
for attachment in message.attachments:
    if not attachment.filename.lower().endswith(self.whitelist):
        rejected_attachments.append(attachment)
    if attachment.filename.lower().endswith('.py'):
        detected_pyfile.append(attachment)
```

Reply:
well yeah, it's twice as expensive if you iterate the same list twice. But I also think it has better readability when you avoid the list comprehensions. Consider for example this line:

```py
len([a for a in message.attachments if a.filename.lower().endswith('.py')]) > 0
```

You'd read this sequentially by saying to yourself "okay, so it's the length of, uh, a list that for every attachment in message.attachment but only if the filename ends with py, a list of items that match this condition, if this list is greater than zero."

Then let's look at the other example (simplified):

```py
for attachment in message.attachments:
    if attachment.filename.lower().endswith('.py'):
        detected_pyfile.append(attachment)
```

For every attachment in message.attachments, if the attachment filename endswith .py, then we append the attachment to this list. It's succinct. It reads almost like an english sentence.

As a rule of thumb, try to keep comprehensions very simple. once you have a comprehension that is complex, just go ahead and refactor it into a loop. And in this case, we have both comprehension complexity, AND a needless second iteration of the same container! it's a no-brainer to refactor.

Hope that's helpful.